DirectShow: VideoCapture returns blank frames for OBS Virtual Camera on Windows

##### System information (version)
- OpenCV => 4.5.1.48 (opencv-python)
- Operating System / Platform => Windows 10 64 Bit


##### Detailed description

[OBS](https://obsproject.com/) installs a DirectShow virtual camera which works fine in Zoom etc. but not in OpenCV via VideoCapture, all frames are either zeros or a bit of garbage at the top of the frame. I also checked it against the [CommandCam](https://github.com/mrihtar/CommandCam) tool which is a simple C++ frame grabber written in a single straight-forward file. No problems there as well. This is what `commandcam.exe /details` shows:

```
  2. OBS Virtual Camera
     Pin: Video Output, 6 caps
       Cap  0: FORMAT_VideoInfo, NV12, 12 bits, 1920x1080 @ 30 fps
       Cap  1: FORMAT_VideoInfo, I420, 12 bits, 1920x1080 @ 30 fps
       Cap  2: FORMAT_VideoInfo, YUY2, 16 bits, 1920x1080 @ 30 fps
       Cap  3: FORMAT_VideoInfo, NV12, 12 bits, 1280x720 @ 20 fps
       Cap  4: FORMAT_VideoInfo, I420, 12 bits, 1280x720 @ 20 fps
       Cap  5: FORMAT_VideoInfo, YUY2, 16 bits, 1280x720 @ 20 fps
```

I tried all pixel formats and resolutions (with the tool) and it worked for all combinations.

##### Steps to reproduce

First, install OBS. The virtual camera always streams a dummy picture.
Then, grab a frame with OpenCV.

```py
def decode_fourcc(fourcc):
    return chr((fourcc & 0xff)) + chr((fourcc & 0xff00) >> 8) + chr((fourcc & 0xff0000) >> 16) + chr((fourcc & 0xff000000) >> 24)

vc = cv2.VideoCapture(2, cv2.CAP_DSHOW)
assert vc.isOpened()

fourcc = decode_fourcc(int(vc.get(cv2.CAP_PROP_FOURCC)))
print(f'fourcc: {fourcc}') # fourcc: }ë6ä  --> with my real webcam this is YUY2

for _ in range(1000):
    ret, frame_captured = vc.read()
    assert ret
    print(frame_captured.min(), frame_captured.max())
    # 0 0
```

I should also mention that I set the environment variable `OPENCV_DSHOW_SAVEGRAPH_FILENAME=graph.grf` and [GraphEdit](https://docs.microsoft.com/en-us/windows/win32/directshow/using-graphedit) cannot open the file. It can open the file when I capture something from my real webcam device however. So somehow the filter graph must be broken.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   <!--
   OpenCV team works with forum.opencv.org, Stack Overflow and other communities
   to discuss problems. Tickets with question without real issue statement will be
   closed.
   -->
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   <!--
   Places to check:
   * OpenCV documentation: https://docs.opencv.org
   * FAQ page: https://github.com/opencv/opencv/wiki/FAQ
   * OpenCV forum: https://forum.opencv.org
   * OpenCV issue tracker: https://github.com/opencv/opencv/issues?q=is%3Aissue
   * Stack Overflow branch: https://stackoverflow.com/questions/tagged/opencv
   -->
 - [x] I updated to latest OpenCV version and the issue is still there
   <!--
   master branch for OpenCV 4.x and 3.4 branch for OpenCV 3.x releases.
   OpenCV team supports only latest release for each branch.
   The ticket is closed, if the problem is not reproduced with modern version.
   -->
 - [x] There is reproducer code and related data files: videos, images, onnx, etc
   <!--
   The best reproducer -- test case for OpenCV that we can add to the library.
   Recommendations for media files and binary files:
   * Try to reproduce the issue with images and videos in opencv_extra repository
     to reduce attachment size
   * Use PNG for images, if you report some CV related bug, but not image reader
     issue
   * Attach the image as archive to the ticket, if you report some reader issue.
     Image hosting services compress images and it breaks the repro code.
   * Provide ONNX file for some public model or ONNX file with with random weights,
     if you report ONNX parsing or handling issue. Architecture details diagram
     from netron tool can be very useful too. See https://lutzroeder.github.io/netron/
   -->

> With OBS 28, my suggestion/workaround is no longer valid, as that plugin won't be updated. It's now completely impossible to use OpenCV with the OBS Virtual Camera.
> 
> Btw here's the error if you turn them on:
> 
> ```
> cv2.error: OpenCV(4.5.5) D:\a\opencv-python\opencv-python\opencv\modules\videoio\src\cap.cpp:365: error: (-2:Unspecified error) could not retrieve channel 0 in function 'cv::VideoCapture::retrieve'
> ```
> 
> Edit: workaround is back, I've updated the plugin: https://github.com/Avasam/obs-virtual-cam/releases

It's the same for me, opencv keeps showing me nothing (well, instead of an array with 0, now it's an array with 128)

--- a/modules/videoio/src/cap_dshow.cpp
+++ b/modules/videoio/src/cap_dshow.cpp
@@ -826,7 +826,8 @@ void videoDevice::setSize(int w, int h){
         {
             videoSize      = w * h;
         }
-        else if (pAmMediaType->subtype == MEDIASUBTYPE_Y16)
+        else if (pAmMediaType->subtype == MEDIASUBTYPE_Y16 ||
+            pAmMediaType->subtype == MEDIASUBTYPE_YUY2)
         {
             videoSize      = w * h * 2;
         }
@@ -1573,6 +1574,12 @@ bool videoInput::getPixels(int id, unsigned char * dstBuffer, bool flipRedAndBlu
                     else {
                         processPixels(src, dst, width, height, flipRedAndBlue, flipImage, 2);
                     }
+                }
+                else if (VDList[id]->pAmMediaType->subtype == MEDIASUBTYPE_YUY2)
+                {
+                    cv::Mat srcMat(height, width, CV_8UC2, src);
+                    cv::Mat dstMat(height, width, CV_8UC3, dst);
+                    cv::cvtColor(srcMat, dstMat, cv::COLOR_YUV2BGR_YUY2);
                 }
                 else
                 {
@@ -2758,6 +2765,14 @@ int videoInput::start(int deviceID, videoDevice *VD){
     if(customSize){
         DebugPrintOut("SETUP: Default Format is set to %ix%i\n", currentWidth, currentHeight);
 
+        if (strcmp("OBS Virtual Camera", VD->nDeviceName) == 0)
+        {
+            // OBS Virtual Camera always returns S_OK on SetFormat(), even if it doesn't support
+            // the actual format. So we have to choose a format that it supports manually, e.g. YUY2.
+            // https://github.com/opencv/opencv/issues/19746#issuecomment-1383056787
+            VD->tryVideoType = MEDIASUBTYPE_YUY2;
+        }
+
         char guidStr[8];
             // try specified format and size
             getMediaSubtypeAsString(VD->tryVideoType, guidStr);
@@ -2869,7 +2884,9 @@ int videoInput::start(int deviceID, videoDevice *VD){
     mt.majortype     = MEDIATYPE_Video;
 
     // Disable format conversion if using 8/16-bit data (e-Con systems)
-    if (checkSingleByteFormat(VD->pAmMediaType->subtype) || (VD->pAmMediaType->subtype == MEDIASUBTYPE_Y16)) {
+    if (checkSingleByteFormat(VD->pAmMediaType->subtype) ||
+        (VD->pAmMediaType->subtype == MEDIASUBTYPE_Y16 || VD->pAmMediaType->subtype == MEDIASUBTYPE_YUY2))
+    {
         DebugPrintOut("SETUP: Not converting frames to RGB.\n");
         mt.subtype = VD->pAmMediaType->subtype;
     }

